Plots

dmpetrov opened the pull request:
Plot metrics.

Related to https://github.com/iterative/dvc/issues/3409

dmpetrov commented:
@jorgeorpinel and @shcheklein for some reason I don't see `dvc plot` & `dvc plot show\diff` in the menu. It looks like I missed something.
Could you please help me?

shcheklein commented:
@dmpetrov yep, update sidebar.json - that should do the trick.

dmpetrov commented:
> sidebar

Thank you @shcheklein ! Forgot about this :)

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,188 @@
+# plot
+
+Contains commands to visualize users data stored in structured files like JSON,
+CSV, TSV: [show](/doc/command-reference/plot/show).
```
```suggestion
Contains commands to visualize
[continuous metrics](/doc/command-reference/plot#continuous-metrics) in
structured files like JSON, CSV, TSV:
[show](/doc/command-reference/plot/show),
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
````diff
@@ -0,0 +1,188 @@
+## Synopsis
+
+```usage
+usage: dvc plot [-h] [-q | -v] {show,diff} ...
+positional arguments:
+  {show,diff}    Use `dvc plot CMD --help` to display command-specific help.
+    show         Plot data from a file
````
```suggestion
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND
    show         Plot data from a file
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,188 @@
+
+## Description
+
+DVC provides a set of commands to visualize
+[continuous metrics](/doc/command-reference/plot#continous-metrics) of machine
+learning experiments in <abbr>DVC projects</abbr>. These metrics are usually
+plotted as AUC curves, loss functions, confusion matrixes and other types of
+plots.
```
```suggestion
DVC provides a set of commands to visualize continuous metrics of machine
learning experiments in. Usual examples of plots are AUC curves, loss
functions, and confusion matrices.
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,188 @@
+can work with these files commited to a repository history, data files
+controlled by DVC or files from workspace.
+
+### Continous metrics
```
```suggestion
### Continuous metrics
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,188 @@
+DVC has two concepts for metrics for representing result of machine learning
+training or data processing:
+
+1. `dvc metrics` to represent scalar numbers such as AUC, true positive rate and
+   others.
+2. `dvc plot` to visualize continuous metrics such as AUC curve, loss function,
+   confusion matrixes and others.
```
Question: maybe we should use "scalar metrics" everywhere i.e. in the dvc metrics commands and other places where they're currently mentioned? Always try to use either scalar or continuous except in placs where metrics are mentioned in an abstract sense.

dmpetrov replied:
> maybe we should use "scalar metrics" everywhere

Yes. It should be a very clear separation.

jorgeorpinel replied:
OK added a checkbox to #1175

dmpetrov commented:
> Please let us know when this is ready for full review.

@jorgeorpinel  thank you for the small changes.

I'm waiting for a bugfix while I can add one more part about the plot templates. It would be great to make review now. I'll incorporate the template later right here or as a separate PR.

dmpetrov commented on content/docs/command-reference/plot/diff.md:
```diff
@@ -0,0 +1,113 @@
+
+## Examples
+
+The difference between a not commited version of the file and the last commited
+one:
```
provided a additional sentence.

dmpetrov commented on content/docs/command-reference/plot/show.md:
````diff
@@ -0,0 +1,130 @@
+## Synopsis
+
+```usage
+usage: dvc plot show [-h] [-q | -v] [-t [TEMPLATE]] [-r RESULT] [--show-json]
````
right

dmpetrov commented:
@jorgeorpinel and @pared pushed all the changes. Please take a look.

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,189 @@
+
+## Description
+
+DVC provides a set of commands to visualize continuous metrics of machine
```
```suggestion
DVC provides a set of commands to visualize _continuous metrics_ of machine
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
````diff
@@ -0,0 +1,189 @@
+37, 0.92302, 0.0299015
+```
+
+In hierarchical file formats such as JSON an array of JSON-objects is expected.
+Plot command can generate visuals for a specified field name or a set of fields
+from the array's object. Like `val_loss` field in the `train` array in this
+example:
````
```suggestion
In JSON files (hierarchical format) a named array of matching JSON objects is
expected. `dvc plot show` can generate visuals for a one, several, or all field
names from the array's objects. For example `val_loss` in the `train` array below:
```

jorgeorpinel commented on content/docs/command-reference/plot/show.md:
```diff
@@ -0,0 +1,130 @@
+# plot show
+
+Generate a [plot](/doc/command-reference/plot) from a file with
+[continuous metrics](/doc/command-reference/plot#continous-metrics).
```
```suggestion
Generate a plot image file from a
[continuous metrics](/doc/command-reference/plot#continous-metrics) file.
```

jorgeorpinel commented on content/docs/command-reference/plot/show.md:
```diff
@@ -0,0 +1,130 @@
+## Description
+
+This command provides a quick way to visualize countinuous metrics such as loss
+functions, AUC curves, confusion matrixes etc. Please read `dvc params` for more
```
```suggestion
functions, AUC curves, confusion matrixes etc. Please read `dvc plot` for more
```

jorgeorpinel commented on content/docs/command-reference/plot/diff.md:
```diff
@@ -0,0 +1,115 @@
+
+Show difference in
+[continuous metrics](/doc/command-reference/plot#continous-metrics) by plotting
+on a single [plot](/doc/command-reference/plot) different versions of metrics
+from the <abbr>DVC repository</abbr> or workspace.
```
I would simplify like this:
```suggestion
different versions in a single image.
```
and leave the other details for the description.

jorgeorpinel commented on content/docs/command-reference/plot/show.md:
```diff
@@ -0,0 +1,130 @@
+                     [datafile]
+
+positional arguments:
+  datafile              Data to be visualized.
```
```suggestion
  datafile              Continuous metrics file to visualize
```
Cc @pared for the core repo PR

p.s. should this term use singular or plural "metric(s)" ?

jorgeorpinel replied:
Moved to iterative/dvc/pull/3577

dmpetrov replied:
metrics - plural in many docs. let's use plural.

jorgeorpinel replied:
This has been applied in the core repo, so we can do it here too 🙂

pared replied:
I don't like `continuous`, my doubts are summarized in my first comment.

jorgeorpinel replied:
Agree, but that seems to be the paragraph in the core repo right now so I'm going to commit this for now. We can update it again when needed.

jorgeorpinel replied:
Ah never mind I see its `Metrics file to visualize.` now. Will update 🙂

jorgeorpinel replied:
Btw extra period 😅

jorgeorpinel replied:
p.s. please see iterative/dvc/pull/3759 @pared

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,189 @@
+
+positional arguments:
+  COMMAND
+    show         Plot data from a file
+    diff         Plot changes between commits in the DVC repository,
+                 or between the last commit and the workspace.
```
```suggestion
    show         Generate a plot image file from a continuous metrics file
    diff         Plot continuous metrics differences between commits in the DVC repository,
                 or between the last commit and the workspace.
```
> May need rewrapping
Cc @pared also for the core repo PR

jorgeorpinel replied:
Moved to iterative/dvc/pull/3577

jorgeorpinel replied:
The definition for show is still pending in the core repo (https://github.com/iterative/dvc/pull/3577#discussion_r416844217)
But the one for diff has been applied ✅

jorgeorpinel replied:
Oh I see this one has been applied. Thanks, resolving.

dmpetrov commented on content/docs/command-reference/plot/index.md:
````diff
@@ -0,0 +1,189 @@
+                 or between the last commit and the workspace.
+```
+
+## Description
+
+DVC provides a set of commands to visualize continuous metrics of machine
+learning experiments in. Usual examples of plots are AUC curves, loss functions,
+and confusion matrices.
````
All points are great! I rearranged this part. Please take a look.

It is still important to keep the scalars vs continuous difference somewhere. I keep it as a next section but rearranged from the plot point of view.

jorgeorpinel replied:
Thanks! I think it's easier to read now 🙂 

> ![image](https://user-images.githubusercontent.com/1477535/80649495-c61f5d00-8a37-11ea-8282-c2235633b016.png)

I agree! Unfortunately I pasted the wrong text in my previous message and so I didn't express myself correctly haha. This is what I meant:

![image](https://user-images.githubusercontent.com/1477535/80649484-bdc72200-8a37-11ea-9001-e45b99e148ba.png)

So to summarize, I think the scalar metrics should be explained in the description of `dvc metrics` cmd ref index, and just mention the word here with a link to there, without explaining. But OK I guess that can be a separate issue (to match dvc metrics with this ref)

jorgeorpinel replied:
Oops I did it again (wrong paste). Please read the comment above in Github ☝️ because I edited it...

jorgeorpinel replied:
p.s. unresolving for myself to remember about "that can be a separate issue (to match dvc metrics with this ref)"

dmpetrov commented:
@jorgeorpinel thank you for the review. Please take a look at the changes.

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,190 @@
+
+## Description
+
+DVC provides a set of commands to visualize _continuous metrics_ of machine
+learning experiments. Usual examples of plots are AUC curves, loss functions,
+and confusion matrices.
```
```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

jorgeorpinel commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,190 @@
+learning experiments. Usual examples of plots are AUC curves, loss functions,
+and confusion matrices.
+
+Continous metrics represents a plot and should be stored as data series in one
+of the supported [file formats](#file-formats). These files are usually created
+by users or generated by user's modeling or data processing code.
```
```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

jorgeorpinel replied:
But how are plots defined? Is there a new option in `dvc run` for these, like -m for metrics? I don't see info about this in any plot ref yet, nor an update on run itself. Does that mean these metrrics are not tracked by DVC like scalar metrics? Probably worth mentioning either way.

dmpetrov replied:
@jorgeorpinel there is no plot metrics in `dvc run` yet. we should introduce one. `-p` is taken by params. What do you think about `-l`\`-L` (pLot)?

jorgeorpinel replied:
I like `-l` (`--plots`) since it comes after `-m` (`--metrics`)

Will there also be a upper case `-L` (`--plot-no-cache`)?

All of this should probably be its own issue in the core repo 😬 I can create it if you'd like, please lmk.

jorgeorpinel commented:
> thank you for the review. Please take a look at the changes

No problem. I resolved all the parts of the review that are now addressed or no longer relevant, and added one more comment. Still haven't reviewed everything in detail (options, examples) is it a good time to do so? And again, I can take this over at some point if needed, lmk guys cc @ivan

shcheklein commented:
@jorgeorpinel wrong `@ ivan` I think :)

what is the status of this?

jorgeorpinel commented:
> wrong @ ivan

I keep doing that...

> what is the status of this?

Some review items are still pending. Are we taking it over?

shcheklein commented:
@jorgeorpinel I trust you on making this call :) don't have enough context on this.

dmpetrov commented:
@shcheklein and @jorgeorpinel I just updated it:
1. made all the asked changes
2. added new plot options
3. small API fixes (output file name)
4. added custom plots section

Please take a look,

pared commented on content/docs/command-reference/plot/diff.md:
```diff
@@ -0,0 +1,130 @@
+# plot diff
+
+Show multiple versions of
+[continuous metrics](/doc/command-reference/plot#continous-metrics) by plotting
```
I wonder whether we want to name them `continuous`. This word applies to functions. What about, for example, confusion matrix? Data for that type of plot is not continuous.

shcheklein replied:
Agreed with @pared . Probably even plain explicit "non-scalar metric" would be better?

dmpetrov replied:
yeah. it will be changed. all the terminology around continuous will be removed in the next iteration.

jorgeorpinel replied:
- [ ] So let's leave "continuous" in this PR so we have a plot cmd ref for now, and update it in a following PR.

pared commented on content/docs/command-reference/plot/index.md:
```diff
@@ -0,0 +1,239 @@
+
+![](/img/plot_auc.svg)
+
+### Difference between continuous and scolar metrics
```
scalar

jorgeorpinel replied:
Adding suggestion for quick commit from GH:
```suggestion
### Difference between continuous and scalar metrics
```
